upup/pkg/fi/cloudup/openstack/cloud.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstack
+
+import (
+	"fmt"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/cloudinstances"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider"
+)
+
+type OpenstackCloud interface {
+	fi.Cloud
+}
+
+type openstackCloud struct {
+}
+
+var _ fi.Cloud = &openstackCloud{}
+
+func NewOpenstackCloud() (OpenstackCloud, error) {
+	return &openstackCloud{}, nil
+}
+
+func (c *openstackCloud) ProviderID() kops.CloudProviderID {
+	return kops.CloudProviderOpenstack
+}
+
+func (c *openstackCloud) DNS() (dnsprovider.Interface, error) {
+	return nil, fmt.Errorf("has not implemented this method")
```

Comment:
Tip: because go errors don't have stack traces, this can be really annoying to track down.  You just see "has not implemented this method" and you don't know which one it was :-)

I normally put the method name in the error ("openstackCloud::DNS not implemented") or I panic or glog.Fatalf, to force a stack-trace :-)

Reply:
thanks for your guidance! good suggestion! I will update it!